src/rules/noAccessStateInSetstateRule.ts

```diff
@@ -0,0 +1,73 @@
+/**
+ * @license
+ * Copyright 2018 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "tslint";
+import { isCallExpression, isPropertyAccessExpression } from "tsutils";
+import * as ts from "typescript";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-access-state-in-setstate",
+        description: "Reports usage of this.state within setState",
+        rationale: Lint.Utils.dedent
+            `Usage of this.state might result in errors when two state calls are \
+            called in batch and thus referencing old state and not the current state.`,
+        options: null,
+        optionsDescription: "",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = "Avoid using this.state in first argument of setState.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    return ts.forEachChild(ctx.sourceFile, callbackForEachChild);
+
+    function callbackForEachChild(node: ts.Node): void {
+        if (!isCallExpression(node)) {
+            return ts.forEachChild(node, callbackForEachChild);
+        }
+
+        const isSetStateCall = node.expression.getText().startsWith("this.setState");
+        if (!isSetStateCall || node.arguments.length === 0) {
+            return ts.forEachChild(node, callbackForEachChild);
+        }
+
+        ts.forEachChild(node.arguments[0], callbackForEachChildInSetStateArgument);
+        return;
+    }
+
+    function callbackForEachChildInSetStateArgument(node: ts.Node): void {
+        if (!isPropertyAccessExpression(node)) {
+            return ts.forEachChild(node, callbackForEachChildInSetStateArgument);
+        }
+
+        if (!node.getText().startsWith("this.state.")) {
```

Comment:
Since `this.state` could also be a umber, boolean or be to some function without accessing it's fields I would suggest to omit the last `.` in the search expression. (You might want to add a test case for that one.)

Reply:
I just realize that I maybe should start the linter with checking if we're even in some kind of React component class.
And yeah, I guess you're right, `this.state` doesn't necessarly need to be an object. I tried to find some source telling you that you shouldn't store some primitive value directly as a state, but couldn't find any, so maybe that isn't even a code smell.. There will be a warning from React though ([this one](https://github.com/facebook/react/blob/master/scripts/error-codes/codes.json#L108)).
But I will transform that check to a RegEx, the one in Line 52 as well :-) Something like `^this\.state\b`

Reply:
Yeah, we should aim for a performant, fast rule, so I'll gladly check with TSLint's guide again. Anything particular you have in mind?

Reply:
I think the current impllementation makes heavy use of the `getText` method, and the guide has some comments regarding it. I think this and checking that the file even can contain/contains `this.setState` might be the most low hanging fruits.

Reply:
Sorry, it took me some time. But now I guess I did improve the performance. I got rid of both ``getText`` calls, and in addition only walk the children of classes.

This linter will still run in every class, ``React.Component`` or not, but as far as I know you'd need a TypedRule to prevent this. But definitely something that could still be improved.

Thanks for the helpful suggestions, sorry for the long delay, and again looking forward to the code review!